File: src/scripts/check_bite_sized_granularity.py
#!/usr/bin/env python3
"""Bite-sized task granularity gate for structural roadmaps (P1.5).

Adopted from `obra/superpowers` `writing-plans/SKILL.md` § Task Structure +
§ No Placeholders (v5.1.0). Complexity-gating is our addition (Council
Round 1, Q4) — only roadmaps tagged `complexity: structural` in frontmatter
are subject to the granularity rules; `complexity: lightweight` skips.

Public API (stdlib-only):

    read_complexity(text)            -> 'structural' | 'lightweight' | None
    scan_placeholders(text)          -> list[Placeholder]
    check_granularity(text)          -> Result(complexity, gated, violations)

`gated` is True only when `complexity == 'structural'`. Violations are
empty when the gate is not active, regardless of placeholder presence.

The CI contract for P1.5 is the pytest harness in
`tests/test_bite_sized_granularity.py`; this module is the test surface.
"""
from __future__ import annotations

import re
from dataclasses import dataclass, field

PLACEHOLDER_PATTERNS: tuple[tuple[str, re.Pattern[str]], ...] = (
    ("angle-placeholder", re.compile(r"<[a-z][a-z0-9 _\-/]*>", re.IGNORECASE)),
    ("todo", re.compile(r"\bTODO\b")),
    ("fixme", re.compile(r"\bFIXME\b")),
    ("xxx", re.compile(r"\bXXX\b")),
    ("tbd", re.compile(r"\btbd\b", re.IGNORECASE)),
    ("triple-question", re.compile(r"\?\?\?")),
)

COMPLEXITY_PAT = re.compile(
    r"^complexity:\s*(lightweight|structural)\s*$", re.MULTILINE
)


@dataclass(frozen=True)
class Placeholder:
    kind: str
    line: int
    text: str


@dataclass
class Result:
    complexity: str | None
    gated: bool
    violations: list[Placeholder] = field(default_factory=list)


def _frontmatter(text: str) -> str:
    if not text.startswith("---\n"):
        return ""
    end = text.find("\n---\n", 4)
    return text[4:end] if end != -1 else ""


def read_complexity(text: str) -> str | None:
    """Return the `complexity:` value from the roadmap frontmatter, or None."""
    fm = _frontmatter(text)
    if not fm:
        return None
    m = COMPLEXITY_PAT.search(fm)
    return m.group(1) if m else None


def scan_placeholders(text: str) -> list[Placeholder]:
    """Return every placeholder hit in task-bullet lines (`- [ ]` / `- [x]`)."""
    hits: list[Placeholder] = []
    for line_no, line in enumerate(text.splitlines(), start=1):
        stripped = line.lstrip()
        if not stripped.startswith(("- [ ]", "- [x]", "- [/]", "- [-]")):
            continue
        for kind, pat in PLACEHOLDER_PATTERNS:
            if pat.search(line):
                hits.append(Placeholder(kind=kind, line=line_no, text=line.rstrip()))
                break
    return hits


def check_granularity(text: str) -> Result:
    """Run the granularity gate.

    Structural roadmaps fail on any placeholder hit in task bullets.
    Lightweight or untagged roadmaps skip the gate (gated=False) and
    return an empty violation list even when placeholders are present.
    """
    complexity = read_complexity(text)
    gated = complexity == "structural"
    if not gated:
        return Result(complexity=complexity, gated=False, violations=[])
    return Result(
        complexity=complexity,
        gated=True,
        violations=scan_placeholders(text),
    )

File: src/scripts/check_bite_sized_granularity.ts
#!/usr/bin/env tsx
/**
 * Bite-sized task granularity gate for structural roadmaps (P1.5).
 *
 * TypeScript twin of `src/scripts/check_bite_sized_granularity.py` (ADR-092,
 * Phase 4 / Wave 4c). This module is a pure library (no CLI / main); the
 * Python module is imported as a test surface. The exported API mirrors the
 * Python public API EXACTLY:
 *
 *     read_complexity(text)   -> 'structural' | 'lightweight' | null
 *     scan_placeholders(text) -> Placeholder[]
 *     check_granularity(text) -> Result(complexity, gated, violations)
 *
 * `gated` is true only when `complexity === 'structural'`. Violations are
 * empty when the gate is not active, regardless of placeholder presence.
 * No behaviour changes — placeholder patterns, frontmatter slice, and the
 * task-bullet prefixes are replicated.
 */

// Placeholder patterns in (kind, regex) order. Mirrors Python re flags:
// IGNORECASE for the angle/tbd patterns. JS uses `i`. The angle pattern
// is global-less here; we use `.test()` per line (re.search semantics).
const PLACEHOLDER_PATTERNS: ReadonlyArray<[string, RegExp]> = [
    ['angle-placeholder', /<[a-z][a-z0-9 _\-/]*>/i],
    ['todo', /\bTODO\b/],
    ['fixme', /\bFIXME\b/],
    ['xxx', /\bXXX\b/],
    ['tbd', /\btbd\b/i],
    ['triple-question', /\?\?\?/],
];

const COMPLEXITY_PAT = /^complexity:\s*(lightweight|structural)\s*$/m;

interface Placeholder {
    kind: string;
    line: number;
    text: string;
}

interface Result {
    complexity: string | null;
    gated: boolean;
    violations: Placeholder[];
}

function _frontmatter(text: string): string {
    if (!text.startsWith('---\n')) {
        return '';
    }
    const end = text.indexOf('\n---\n', 4);
    return end !== -1 ? text.slice(4, end) : '';
}

function read_complexity(text: string): string | null {
    const fm = _frontmatter(text);
    if (fm === '') {
        return null;
    }
    const m = COMPLEXITY_PAT.exec(fm);
    return m ? m[1]! : null;
}

const _TASK_PREFIXES = ['- [ ]', '- [x]', '- [/]', '- [-]'];

/** Mirror Python `str.lstrip()` — strip leading whitespace (incl. tabs). */
function _lstrip(s: string): string {
    return s.replace(/^\s+/, '');
}

/** Mirror Python `str.rstrip()` — strip trailing whitespace. */
function _rstrip(s: string): string {
    return s.replace(/\s+$/, '');
}

function scan_placeholders(text: string): Placeholder[] {
    const hits: Placeholder[] = [];
    const lines = text.split('\n');
    for (let idx = 0; idx < lines.length; idx++) {
        const line_no = idx + 1;
        const line = lines[idx]!;
        const stripped = _lstrip(line);
        if (!_TASK_PREFIXES.some((p) => stripped.startsWith(p))) {
            continue;
        }
        for (const [kind, pat] of PLACEHOLDER_PATTERNS) {
            if (pat.test(line)) {
                hits.push({ kind, line: line_no, text: _rstrip(line) });
                break;
            }
        }
    }
    return hits;
}

function check_granularity(text: string): Result {
    const complexity = read_complexity(text);
    const gated = complexity === 'structural';
    if (!gated) {
        return { complexity, gated: false, violations: [] };
    }
    return {
        complexity,
        gated: true,
        violations: scan_placeholders(text),
    };
}

export {
    type Placeholder,
    type Result,
    PLACEHOLDER_PATTERNS,
    COMPLEXITY_PAT,
    read_complexity,
    scan_placeholders,
    check_granularity,
};
